File: src/backend/IAW.Vdf.Authoring/Paraphrase/RoundTripParaphraser.cs
using IAW.Vdf.Abstractions.Conditions;
using IAW.Vdf.Abstractions.Outcomes;
using IAW.Vdf.Abstractions.Rules;

namespace IAW.Vdf.Authoring.Paraphrase;

/// <summary>
/// Produces deterministic, human-readable English sentences from a <see cref="RuleDefinition"/>.
/// The output is stable: calling <see cref="Paraphrase"/> twice with the same input always yields
/// the same string.
/// </summary>
public sealed class RoundTripParaphraser
{
    /// <summary>
    /// Converts a <see cref="RuleDefinition"/> to a descriptive English sentence.
    /// </summary>
    /// <param name="rule">The rule to paraphrase.</param>
    /// <returns>A deterministic English description of the rule's semantics.</returns>
    public string Paraphrase(RuleDefinition rule)
    {
        // Derivation rules have no Assert — they produce an outcome based on AppliesWhen alone.
        if (rule.Assert is null)
        {
            // "When {appliesWhen}, {outcome}."
            // or for no appliesWhen: "Always: {outcome}."
            if (rule.AppliesWhen is not null)
            {
                var when = RenderCondition(rule.AppliesWhen);
                var outcome = RenderOutcome(rule.OnFailure);
                return $"When {when}, {outcome}.";
            }
            else
            {
                var outcome = RenderOutcome(rule.OnFailure);
                return $"Always: {outcome}.";
            }
        }

        // Validation / Route rules have an Assert.
        var assertPhrase = RenderCondition(rule.Assert);

        if (rule.AppliesWhen is not null)
        {
            var whenPhrase = RenderCondition(rule.AppliesWhen);
            var recoveryPhrase = rule.Recover is not null
                ? RenderRecovery(rule.Recover) + "; if unresolved, "
                : "";
            var failurePhrase = RenderOutcome(rule.OnFailure);

            return $"For orders where {whenPhrase}, require {assertPhrase} to be present; if absent, {recoveryPhrase}{failurePhrase}.";
        }
        else
        {
            var recoveryPhrase = rule.Recover is not null
                ? RenderRecovery(rule.Recover) + ". If unresolved, "
                : "";
            var failurePhrase = RenderOutcome(rule.OnFailure);

            return $"Require {assertPhrase} to be present; {recoveryPhrase}{failurePhrase}.";
        }
    }

    // ── Condition rendering ──────────────────────────────────────────────────────────────────────

    private static string RenderCondition(ICondition condition) =>
        condition switch
        {
            LeafCondition leaf => RenderLeaf(leaf),
            GroupCondition group => RenderGroup(group),
            _ => condition.ToString() ?? "unknown condition"
        };

    private static string RenderLeaf(LeafCondition leaf)
    {
        var subject = leaf.Subject;
        var prefix = leaf.Quantifier switch
        {
            Quantifier.Any => $"any {subject}",
            Quantifier.Every => $"every {subject}",
            _ => subject
        };

        return leaf.Operator switch
        {
            OperatorKind.IsPresent => $"{prefix} is present",
            OperatorKind.IsAbsent => $"{prefix} is absent",
            OperatorKind.Equals => $"{prefix} equals {RenderValue(leaf)}",
            OperatorKind.NotEquals => $"{prefix} does not equal {RenderValue(leaf)}",
            OperatorKind.InSet => $"{prefix} is in {RenderRef(leaf)}",
            OperatorKind.NotInSet => $"{prefix} is not in {RenderRef(leaf)}",
            OperatorKind.GreaterThan => $"{prefix} is greater than {RenderRef(leaf)}",
            OperatorKind.LessThan => $"{prefix} is less than {RenderRef(leaf)}",
            OperatorKind.GreaterOrEqual => $"{prefix} is greater than or equal to {RenderRef(leaf)}",
            OperatorKind.LessOrEqual => $"{prefix} is less than or equal to {RenderRef(leaf)}",
            OperatorKind.WithinRange => $"{prefix} is within range {RenderRef(leaf)}",
            OperatorKind.Matches => $"{prefix} matches {RenderRef(leaf)}",
            OperatorKind.IsCompatibleWith => $"{prefix} is compatible with {RenderRef(leaf)}",
            OperatorKind.IsEligibleFor => $"{prefix} is eligible for {RenderRef(leaf)}",
            OperatorKind.Exists => $"{prefix} exists",
            _ => $"{prefix} {leaf.Operator} {RenderValue(leaf)}"
        };
    }

    private static string RenderGroup(GroupCondition group)
    {
        var children = group.Conditions.Select(RenderCondition).ToList();
        return group.LogicalOp switch
        {
            LogicalOperator.All => string.Join(" and ", children),
            LogicalOperator.Any => string.Join(" or ", children),
            LogicalOperator.Not when children.Count == 1 => $"not ({children[0]})",
            _ => $"({string.Join(", ", children)})"
        };
    }

    private static string RenderValue(LeafCondition leaf)
    {
        if (leaf.Value is not null)
            return leaf.Value.ToJsonString().Trim('"');
        if (leaf.Reference is not null)
            return leaf.Reference;
        return "null";
    }

    private static string RenderRef(LeafCondition leaf) =>
        leaf.Reference ?? (leaf.Value is not null ? leaf.Value.ToJsonString().Trim('"') : "null");

    // ── Outcome rendering ───────────────────────────────────────────────────────────────────────

    private static string RenderOutcome(Outcome outcome)
    {
        var scope = outcome.Scope ?? "order";
        var reason = outcome.Reason;
        var reasonSuffix = !string.IsNullOrWhiteSpace(reason) ? $": {reason}" : "";

        return outcome.Type switch
        {
            OutcomeType.Continue => "proceed",
            OutcomeType.Suppressed => "suppress the hold (resolved by recovery)",

            OutcomeType.CompleteHold =>
                $"place a complete problem hold on the {scope}{reasonSuffix}",

            OutcomeType.PartialHold =>
                $"place a partial hold on the {scope}{reasonSuffix}",

            OutcomeType.Warning =>
                $"raise a warning on the {scope}{reasonSuffix}",

            OutcomeType.ComplianceAlert =>
                BuildComplianceAlert(outcome),

            OutcomeType.RouteToReview =>
                BuildRouteToReview(outcome),

            OutcomeType.RouteToQueue =>
                BuildRouteToQueue(outcome),

            OutcomeType.Escalate =>
                $"escalate the {scope}{reasonSuffix}",

            OutcomeType.SetValue =>
                BuildSetValue(outcome),

            OutcomeType.ApplyDefault =>
                BuildApplyDefault(outcome),

            OutcomeType.CalculateValue =>
                BuildCalculateValue(outcome),

            OutcomeType.CreatePlaceholder =>
                BuildCreatePlaceholder(outcome),

            OutcomeType.CreateIncident =>
                $"create an incident for the {scope}{reasonSuffix}",

            OutcomeType.CreateTask =>
                $"create a task for the {scope}{reasonSuffix}",

            OutcomeType.PreventAction =>
                BuildPreventAction(outcome),

            OutcomeType.AllowAction =>
                BuildAllowAction(outcome),

            _ => outcome.Type.ToString()
        };
    }

    private static string BuildComplianceAlert(Outcome outcome)
    {
        var scope = outcome.Scope ?? "order";
        var severity = outcome.Severity;
        var article = string.Equals(severity, "informational", StringComparison.OrdinalIgnoreCase)
            ? "an informational"
            : "a";
        return $"raise {article} compliance alert on the {scope}";
    }

    private static string BuildRouteToReview(Outcome outcome)
    {
        var scope = outcome.Scope ?? "order";
        outcome.Parameters.TryGetValue("Destination", out var dest);
        var destination = dest?.ToString() ?? "review queue";
        return $"route the {scope} to {destination} for review";
    }

    private static string BuildRouteToQueue(Outcome outcome)
    {
        var scope = outcome.Scope ?? "order";
        outcome.Parameters.TryGetValue("Queue", out var queue);
        var queueName = queue?.ToString() ?? "queue";
        return $"route the {scope} to {queueName}";
    }

    private static string BuildSetValue(Outcome outcome)
    {
        outcome.Parameters.TryGetValue("Target", out var target);
        outcome.Parameters.TryGetValue("Value", out var value);
        return $"set {target ?? "target"} to '{value ?? "value"}'";
    }

    private static string BuildApplyDefault(Outcome outcome)
    {
        outcome.Parameters.TryGetValue("Target", out var target);
        return $"apply the default value to {target ?? "target"}";
    }

    private static string BuildCalculateValue(Outcome outcome)
    {
        outcome.Parameters.TryGetValue("Target", out var target);
        return $"calculate the value for {target ?? "target"}";
    }

    private static string BuildCreatePlaceholder(Outcome outcome)
    {
        outcome.Parameters.TryGetValue("SpecimenType", out var specimenType);
        return $"create a placeholder {specimenType ?? "specimen"} specimen";
    }

    private static string BuildPreventAction(Outcome outcome)
    {
        outcome.Parameters.TryGetValue("Action", out var action);
        return $"prevent the '{action ?? "action"}' action";
    }

    private static string BuildAllowAction(Outcome outcome)
    {
        outcome.Parameters.TryGetValue("Action", out var action);
        return $"allow the '{action ?? "action"}' action";
    }

    // ── Recovery rendering ──────────────────────────────────────────────────────────────────────

    private static string RenderRecovery(RecoveryStrategy recover)
    {
        if (string.Equals(recover.Strategy, RecoveryStrategy.ApplyDefault, StringComparison.OrdinalIgnoreCase))
        {
            recover.Parameters.TryGetValue("Target", out var target);
            recover.Parameters.TryGetValue("Reference", out var reference);
            var targetStr = target?.ToString() ?? "target";
            var refStr = reference?.ToString() ?? "reference";
            return $"first attempt to apply default value to {targetStr} from {refStr}";
        }

        if (string.Equals(recover.Strategy, RecoveryStrategy.FindAlternateSpecimen, StringComparison.OrdinalIgnoreCase))
        {
            return "first search for an alternate compatible specimen";
        }

        return $"first attempt recovery via '{recover.Strategy}'";
    }
}

File: src/server/src/authoring/round-trip-paraphraser.ts
/**
 * Deterministic structure → English back-translation of a {@link RuleDefinition}.
 *
 * A faithful port of {@link ../../backend/IAW.Vdf.Authoring/Paraphrase/RoundTripParaphraser.cs}:
 * the output is stable (same rule → same sentence) and covers every operator,
 * quantifier, outcome type, and recovery strategy. Derivation rules (no Assert)
 * render as "When …, …"; validation/route rules render as "require … to be present".
 */

import {
  Condition,
  GroupCondition,
  LeafCondition,
  Outcome,
  RecoveryStrategy,
  RecoveryStrategyName,
  RuleDefinition,
} from '../vdf/types';

/** Produces a deterministic English description of a rule's semantics. */
export class RoundTripParaphraser {
  paraphrase(rule: RuleDefinition): string {
    // Derivation rules have no Assert — they produce an outcome on AppliesWhen alone.
    if (rule.assert === undefined) {
      if (rule.appliesWhen !== undefined) {
        const when = renderCondition(rule.appliesWhen);
        const outcome = renderOutcome(rule.onFailure);
        return `When ${when}, ${outcome}.`;
      }
      return `Always: ${renderOutcome(rule.onFailure)}.`;
    }

    const assertPhrase = renderCondition(rule.assert);

    if (rule.appliesWhen !== undefined) {
      const whenPhrase = renderCondition(rule.appliesWhen);
      const recoveryPhrase = rule.recover
        ? `${renderRecovery(rule.recover)}; if unresolved, `
        : '';
      const failurePhrase = renderOutcome(rule.onFailure);
      return `For orders where ${whenPhrase}, require ${assertPhrase} to be present; if absent, ${recoveryPhrase}${failurePhrase}.`;
    }

    const recoveryPhrase = rule.recover
      ? `${renderRecovery(rule.recover)}. If unresolved, `
      : '';
    const failurePhrase = renderOutcome(rule.onFailure);
    return `Require ${assertPhrase} to be present; ${recoveryPhrase}${failurePhrase}.`;
  }
}

// ── Condition rendering ───────────────────────────────────────────────────────

function renderCondition(condition: Condition): string {
  return condition.type === 'leaf'
    ? renderLeaf(condition)
    : renderGroup(condition);
}

function renderLeaf(leaf: LeafCondition): string {
  const subject = leaf.subject;
  const quantifier = leaf.quantifier ?? 'This';
  const prefix =
    quantifier === 'Any'
      ? `any ${subject}`
      : quantifier === 'Every'
        ? `every ${subject}`
        : subject;

  switch (leaf.operator) {
    case 'IsPresent':
      return `${prefix} is present`;
    case 'IsAbsent':
      return `${prefix} is absent`;
    case 'Equals':
      return `${prefix} equals ${renderValue(leaf)}`;
    case 'NotEquals':
      return `${prefix} does not equal ${renderValue(leaf)}`;
    case 'InSet':
      return `${prefix} is in ${renderRef(leaf)}`;
    case 'NotInSet':
      return `${prefix} is not in ${renderRef(leaf)}`;
    case 'GreaterThan':
      return `${prefix} is greater than ${renderRef(leaf)}`;
    case 'LessThan':
      return `${prefix} is less than ${renderRef(leaf)}`;
    case 'GreaterOrEqual':
      return `${prefix} is greater than or equal to ${renderRef(leaf)}`;
    case 'LessOrEqual':
      return `${prefix} is less than or equal to ${renderRef(leaf)}`;
    case 'WithinRange':
      return `${prefix} is within range ${renderRef(leaf)}`;
    case 'Matches':
      return `${prefix} matches ${renderRef(leaf)}`;
    case 'IsCompatibleWith':
      return `${prefix} is compatible with ${renderRef(leaf)}`;
    case 'IsEligibleFor':
      return `${prefix} is eligible for ${renderRef(leaf)}`;
    case 'Exists':
      return `${prefix} exists`;
    default:
      return `${prefix} ${String(leaf.operator)} ${renderValue(leaf)}`;
  }
}

function renderGroup(group: GroupCondition): string {
  const children = group.conditions.map(renderCondition);
  switch (group.logicalOp) {
    case 'All':
      return children.join(' and ');
    case 'Any':
      return children.join(' or ');
    case 'Not':
      return children.length === 1
        ? `not (${children[0]})`
        : `(${children.join(', ')})`;
    default:
      return `(${children.join(', ')})`;
  }
}

function renderValue(leaf: LeafCondition): string {
  if (leaf.value !== undefined && leaf.value !== null) {
    return stripQuotes(leaf.value);
  }
  if (leaf.reference !== undefined) {
    return leaf.reference;
  }
  return 'null';
}

function renderRef(leaf: LeafCondition): string {
  if (leaf.reference !== undefined) {
    return leaf.reference;
  }
  return leaf.value !== undefined && leaf.value !== null
    ? stripQuotes(leaf.value)
    : 'null';
}

function stripQuotes(value: unknown): string {
  if (typeof value === 'string') {
    return value;
  }
  return JSON.stringify(value);
}

// ── Outcome rendering ─────────────────────────────────────────────────────────

function renderOutcome(outcome: Outcome): string {
  const scope = outcome.scope ?? 'order';
  const reasonSuffix =
    outcome.reason && outcome.reason.trim() !== '' ? `: ${outcome.reason}` : '';

  switch (outcome.type) {
    case 'Continue':
      return 'proceed';
    case 'Suppressed':
      return 'suppress the hold (resolved by recovery)';
    case 'CompleteHold':
      return `place a complete problem hold on the ${scope}${reasonSuffix}`;
    case 'PartialHold':
      return `place a partial hold on the ${scope}${reasonSuffix}`;
    case 'Warning':
      return `raise a warning on the ${scope}${reasonSuffix}`;
    case 'ComplianceAlert':
      return buildComplianceAlert(outcome);
    case 'RouteToReview':
      return buildRouteToReview(outcome);
    case 'RouteToQueue':
      return buildRouteToQueue(outcome);
    case 'Escalate':
      return `escalate the ${scope}${reasonSuffix}`;
    case 'SetValue':
      return buildSetValue(outcome);
    case 'ApplyDefault':
      return `apply the default value to ${paramOr(outcome, 'Target', 'target')}`;
    case 'CalculateValue':
      return `calculate the value for ${paramOr(outcome, 'Target', 'target')}`;
    case 'CreatePlaceholder':
      return `create a placeholder ${paramOr(outcome, 'SpecimenType', 'specimen')} specimen`;
    case 'CreateIncident':
      return `create an incident for the ${scope}${reasonSuffix}`;
    case 'CreateTask':
      return `create a task for the ${scope}${reasonSuffix}`;
    case 'PreventAction':
      return `prevent the '${paramOr(outcome, 'Action', 'action')}' action`;
    case 'AllowAction':
      return `allow the '${paramOr(outcome, 'Action', 'action')}' action`;
    default:
      return String(outcome.type);
  }
}

function buildComplianceAlert(outcome: Outcome): string {
  const scope = outcome.scope ?? 'order';
  const article =
    (outcome.severity ?? '').toLowerCase() === 'informational'
      ? 'an informational'
      : 'a';
  return `raise ${article} compliance alert on the ${scope}`;
}

function buildRouteToReview(outcome: Outcome): string {
  const scope = outcome.scope ?? 'order';
  return `route the ${scope} to ${paramOr(outcome, 'Destination', 'review queue')} for review`;
}

function buildRouteToQueue(outcome: Outcome): string {
  const scope = outcome.scope ?? 'order';
  return `route the ${scope} to ${paramOr(outcome, 'Queue', 'queue')}`;
}

function buildSetValue(outcome: Outcome): string {
  const target = paramOr(outcome, 'Target', 'target');
  const value = paramOr(outcome, 'Value', 'value');
  return `set ${target} to '${value}'`;
}

/** Returns the string-coerced parameter value, or a fallback when absent/blank. */
function paramOr(outcome: Outcome, key: string, fallback: string): string {
  const value = outcome.parameters[key];
  if (value === undefined || value === null) {
    return fallback;
  }
  return typeof value === 'string' ? value : JSON.stringify(value);
}

// ── Recovery rendering ────────────────────────────────────────────────────────

function renderRecovery(recover: RecoveryStrategy): string {
  if (
    recover.strategy.toLowerCase() ===
    RecoveryStrategyName.ApplyDefault.toLowerCase()
  ) {
    const target = stringParam(recover.parameters['Target'], 'target');
    const reference = stringParam(recover.parameters['Reference'], 'reference');
    return `first attempt to apply default value to ${target} from ${reference}`;
  }
  if (
    recover.strategy.toLowerCase() ===
    RecoveryStrategyName.FindAlternateSpecimen.toLowerCase()
  ) {
    return 'first search for an alternate compatible specimen';
  }
  return `first attempt recovery via '${recover.strategy}'`;
}

function stringParam(value: unknown, fallback: string): string {
  if (value === undefined || value === null) {
    return fallback;
  }
  return typeof value === 'string' ? value : JSON.stringify(value);
}
